src/translate.rs

```diff
@@ -1162,38 +1180,77 @@ impl FromGlib<i32> for Option<u64> {
     }
 }
 
-/// Translate from a pointer type without taking ownership, transfer: none.
+/// Translate from a pointer type which is annotated with `transfer none`.
+/// The resulting value is referenced at least once, by the bindings.
+///
+/// This is suitable for floating references, which become strong references.
+/// It is also suitable for acquiring non-gobject values, like `gchar*`.
+///
+/// The implementation of this trait should acquire a reference to the value
+/// in a way appropriate to the type, e.g. by referencing or copying.
+/// Values obtained using this trait must be properly released on `drop()`
+/// by the implementing type.
+///
+/// For more information, refer to module level documentation.
 pub trait FromGlibPtrNone<P: Ptr>: Sized {
     unsafe fn from_glib_none(ptr: P) -> Self;
 }
 
-/// Translate from a pointer type taking ownership, transfer: full.
+/// Translate from a pointer type which is annotated with `transfer full`.
+/// This transfers the ownership of the value to the Rust side.
+///
+/// Because ownership can only be transferred if something is already referenced,
+/// this is unsuitable for floating references.
+///
+/// The implementation of this trait should not acquire any references to the value.
+/// Values obtained using this trait must be properly released on `drop()`
+/// by the implementing type.
+///
+/// For more information, refer to module level documentation.
 pub trait FromGlibPtrFull<P: Ptr>: Sized {
     unsafe fn from_glib_full(ptr: P) -> Self;
 }
 
-/// Translate from a pointer type by borrowing. Don't increase the refcount
+/// Translate from a pointer type by borrowing, without affecting the refcount.
+///
+/// The purpose of this trait is to access values inside callbacks
+/// without changing their reference status.
+/// The obtained borrow must not be accessed outside of the scope of the callback.
+/// Care must be taken not to change the reference count of the borrowed value,
+/// and not to let any called procedure to store a reference.
+/// In particular, safe Rust code should never obtain a mutable borrow.
```

Comment:
Rust code must never have access to the value itself, only to immutable/shared references to it

Reply:
What I (think I) understand:
- no procedure may obtain the value via move, because the lifetime of the underlying pointer is constrained to the scope of the call (but not enforced by the type signature)
- no procedure may be allowed to store a Rust `&` reference for the same reasons, with the same solution

What I don't understand is what's wrong with mutable references being passed around. They can't be stored anyway due to the above, and mutating callback params in C code is the usual thing. It's also not about Glib references, because clone() creates them and is allowed.

Reply:
If you have a mutable reference then you can move (in safe code!) the value out of it via `mem::replace()`. This would then allow to have the value outside the scope of the call.

Reply:
Addressed as "Safe Rust code must never obtain a mutable Rust reference."